flake.nix

```diff
@@ -0,0 +1,177 @@
+            shellHook = ''
+              export -f setuptoolsShellHook runHook _eval _callImplicitHook
+            '';
+          };
+    in
+    {
+      formatter.${system} = pkgs.nixpkgs-fmt;
+
+      packages.${system} = rec {
+        default = diffkemp-llvm16;
+
+        diffkemp-llvm16 = mkDiffkemp pkgs.llvmPackages_16;
+        diffkemp-llvm15 = mkDiffkemp pkgs.llvmPackages_15;
+        diffkemp-llvm14 = mkDiffkemp pkgs.llvmPackages_14;
+        diffkemp-llvm13 = mkDiffkemp pkgs.llvmPackages_13;
+        diffkemp-llvm12 = mkDiffkemp pkgs.llvmPackages_12;
+        diffkemp-llvm11 = mkDiffkemp pkgs.llvmPackages_11;
+        diffkemp-llvm10 = mkDiffkemp pkgs.llvmPackages_10;
+        diffkemp-llvm9 = mkDiffkemp pkgs.llvmPackages_9;
+      };
+
+      devShells.${system} = rec {
+        default = diffkemp-llvm16;
+
+        diffkemp-llvm16 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm16;
+        diffkemp-llvm15 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm15;
+        diffkemp-llvm14 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm14;
+        diffkemp-llvm13 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm13;
+        diffkemp-llvm12 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm12;
+        diffkemp-llvm11 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm11;
+        diffkemp-llvm10 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm10;
+        diffkemp-llvm9 = mkDiffkempDevShell self.packages.${system}.diffkemp-llvm9;
+
+        # Environment for downloading and preparing test kernels (RHEL 7 and 8).
+        # Contains 2 changes from the default env necessary for RHEL 7:
+        #  - gcc 7
+        #  - make 3.81
+        # It should be sufficient to use this to download the kernels (with
+        # rhel-kernel-get) and the tests can be run in the default dev shell.
+        test-kernel-buildenv =
+          with pkgs;
+          let
+            oldmake = import
+              (builtins.fetchTarball {
+                url = "https://github.com/NixOS/nixpkgs/archive/92487043aef07f620034af9caa566adecd4a252b.tar.gz";
+                sha256 = "00fgvpj0aqmq45xmmiqr2kvdir6zigyasx130rp96hf35mab1n8c";
+              })
+              { inherit system; };
+            gnumake381 = oldmake.gnumake381;
+          in
+
+          gcc7Stdenv.mkDerivation {
+            # gcc7Stdenv provides GCC 7, however it doesn't provide the mkShell
+            # command so we need to use mkDerivation and clear phases (that's what
+            # mkShell does).
+            name = "test-kernel-buildenv";
+            phases = [ ];
+
+            nativeBuildInputs = lib.lists.remove gcc default.nativeBuildInputs;
+
+            buildInputs = lib.lists.remove gnumake default.buildInputs ++ [
+              gnumake381
+            ];
+
+            propagatedBuildInputs = default.propagatedBuildInputs;
+
+            dontUseSetuptoolsShellHook = true;
+          };
+      };
+    };
```

Comment:
I was playing with Nix a bit and found out that we could define the set of all development environments like this, and we wouldn't have to update multiple lines with each new LLVM version:
```suggestion
    in {
      formatter.${system} = pkgs.nixpkgs-fmt;

      packages.${system} = (builtins.listToAttrs (map (ver: {
        name = "diffkemp-llvm${ver}";
        value = mkDiffkemp pkgs."llvmPackages_${ver}";
      }) llvmVersionStrings)) // {
        default = mkDiffkemp pkgs."llvmPackages_${llvmLastVersionString}";
      };

      devShells.${system} = (builtins.listToAttrs (map (ver: {
        name = "diffkemp-llvm${ver}";
        value =
          mkDiffkempDevShell self.packages.${system}."diffkemp-llvm${ver}";
      }) llvmVersionStrings)) // {
        default = mkDiffkempDevShell
          self.packages.${system}."diffkemp-llvm${llvmLastVersionString}";
      } // {

        # Environment for downloading and preparing test kernels (RHEL 7 and 8).
        # Contains 2 changes from the default env necessary for RHEL 7:
        #  - gcc 7
        #  - make 3.81
        # It should be sufficient to use this to download the kernels (with
        # rhel-kernel-get) and the tests can be run in the default dev shell.
        test-kernel-buildenv = with pkgs;
          let
            oldmake = import (builtins.fetchTarball {
              url =
                "https://github.com/NixOS/nixpkgs/archive/92487043aef07f620034af9caa566adecd4a252b.tar.gz";
              sha256 = "00fgvpj0aqmq45xmmiqr2kvdir6zigyasx130rp96hf35mab1n8c";
            }) { inherit system; };
            gnumake381 = oldmake.gnumake381;

          in gcc7Stdenv.mkDerivation {
            # gcc7Stdenv provides GCC 7, however it doesn't provide the mkShell
            # command so we need to use mkDerivation and clear phases (that's what
            # mkShell does).
            name = "test-kernel-buildenv";
            phases = [ ];

            nativeBuildInputs = lib.lists.remove gcc default.nativeBuildInputs;

            buildInputs = lib.lists.remove gnumake default.buildInputs
              ++ [ gnumake381 ];

            propagatedBuildInputs = default.propagatedBuildInputs;

            dontUseSetuptoolsShellHook = true;
          };
      };
    };
}
```
Of course, we would have to define some variables:
```nix
llvmVersions = pkgs.lib.lists.range 9 16;
llvmVersionStrings = map toString llvmVersions;
llvmLastVersionString = pkgs.lib.lists.last llvmVersionStrings;
```
With each new LLVM version being packaged in Nix, we would only update the range in `llvmVersions`.

On the other hand, when I started playing with this, I thought the end result would look much cleaner than it does :smile: So feel free to stick with the original version if you don't like this.

Reply:
This actually looks quite nice (and easier to maintain when we add new LLVM versions), thanks! Let me give this a try.

Reply:
I also think, that this looks nicer and more maintainable. I was thinking and some packages in `nixpkgs` are using some kind of "latest" version variable, perhaps we also could use something like that here?

```nix
llvmLatestVersion = 16;
llvmVersions = pkgs.lib.lists.range 9 llvmLatestVersion;
llvmVersionStrings = map toString llvmVersions;
```

And if I am correct (but I am not sure at the moment), then we should be able to use this `llvmLatestVersion` in the format strings without casting.

I have not tried running it, so its more of an idea/suggestion.

Reply:
Thanks, I've implemented this (with some modifications).